### DISCUSSION TITLE
How To I check user before creating?

### DISCUSSION THREAD

**Discussion by pgswow** [pocketbase/pocketbase#1018] [Q&A] [answered]
Hello! i use sveltekit, in one query i autorize user with another auth database (i use LDAP).
How to i check, if user created - i go update profile user and and go authWithPassword().
If user not found - i create user with collection('users').create({ username: username, password: password, passwordConfirm: password }), and go to auth.

**Comment by ganigeorgiev** [comment, marked as answer]
This is difficult to generalize without context what user data you have after the LDAP authentication and can be done several ways depending whether you use PocketBase as framework or not.

I assume that you use the prebuilt v0.8.0-rc binary and that you have access to the username. In this case to check whether a user exist you could do something like:

```js
import PocketBase from "pocketbase";

const pb = new PocketBase("http://127.0.0.1:8090");

// (optionally) authenticate as admin in order to ignore the List API rule
await pb.admins.authWithPassword("test@example.com", "1234567890");

const user = await pb.collection("users").getFirstListItem(`username="${YOUR_USERNAME}"`);
```

**Reply by pgswow** [reply]
Yes, I think this is what you need, but there is still a question with the API rules. At the time of verification, the user is not yet registered and authenticated. How to build API rules correctly. I'm sorry, I just recently started to study PocketBase, its a very cool!
P.S. I use 0.8.0-rc2

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand _" At the time of verification, the user is not yet registered and authenticated"_.

You don't need to worry about API rules if you authenticate as admin first (as shown in the example). API rules don't apply to admins and they have access to everything.

**Reply by pgswow** [reply]
oh sorry, yes! need autentificate as admin =) i chek this =)

**Reply by pgswow** [reply]
It turns out that if there is no registered user on the server that we are looking for, then we will not receive anything in the user variable, but will go straight to the catch block.
It turns out that in this case it is necessary to describe the necessary logic for creating a user in the catch block, right?

Also, when authorizing under an administrator account, an admin cookie is set for the time of user verification, how safe is this?

**Reply by ganigeorgiev** [reply]
@pgswow

> It turns out that in this case it is necessary to describe the necessary logic for creating a user in the catch block, right?

Yes, getFirstListItem returns an error if no user is found (similar to getOne; it is also mentioned in the [jsdoc of the method](https://github.com/pocketbase/js-sdk/blob/rc/src/services/utils/CrudService.ts#L29-L43) but I'll see if I can state it more clearly in the documentation somewhere else).

> Also, when authorizing under an administrator account, an admin cookie is set for the time of user verification, how safe is this?

It is hard to tell without any code sample how you initialize and interact with the SDK. The JS SDK by default doesn't use cookies and the logged state is kept only during runtime (or for the lifetime of the request if you create the client for each node server request).
If you are loading and setting cookies based on the the the [example SSR guide](https://github.com/pocketbase/js-sdk/tree/rc#ssr-integration), then Yes it is possible to "leak" the authenticated state to the client and to avoid it you can create a different client instance or have a check for the AuthStore model whether it is a record (eg. `pb.authStore.model instanceof Record`).

**Reply by pgswow** [reply]
got it thanks, about setting cookies, it makes a hooks.server.js (svelte kit), that's my problem, sorry for that, thanks for the reply!

**Reply by hossxosc8107** [reply]
> This is difficult to generalize without context what user data you have after the LDAP authentication and can be done several ways depending whether you use PocketBase as framework or not.
> 
> I assume that you use the prebuilt v0.8.0-rc binary and that you have access to the username. In this case to check whether a user exist you could do something like:
> 
> ```js
> import PocketBase from "pocketbase";
> 
> const pb = new PocketBase("http://127.0.0.1:8090");
> 
> // (optionally) authenticate as admin in order to ignore the List API rule
> await pb.admins.authWithPassword("test@example.com", "1234567890");
> 
> const user = await pb.collection("users").getFirstListItem(`username="${YOUR_USERNAME}"`);
> ```

This need to open list/search action api rule, may cause the visitor can use js sdk to load all the user data, it's a System Vulnerabilities. User's email, name, phone number will be leak

**Reply by ganigeorgiev** [reply]
@hossxosc8107 I'm not sure that I understand what do you mean. Admins can access and modify everything. Regular users can view other users email only in case the `emailVisibility` flag is set to true. If that's not what you are asking, please file a new discussion with more details since this one was before v0.8.0 release and a lot of things has changed since then.

**Reply by hossxosc8107** [reply]
> @hossxosc8107 I'm not sure that I understand what do you mean. Admins can access and modify everything. Regular users can view other users email only in case the `emailVisibility` flag is set to true. If that's not what you are asking, please file a new discussion with more details since this one was before v0.8.0 release and a lot of things has changed since then.

in fact pb js sdk has a way to check whether username exist with email address

      const data = {
            username: username,
            password: password,
            passwordConfirm: password2    
          }
     const createdUser = await pb.collection('users').create(data)
          window.location.href = '/login'
          //await login()
        } catch (err) {
          console.error(err)
          console.log(err.data)
    
          if (err.data.data.username.code){
            console.log("user already exist")
            //put some code then return        
          } 
          if (err.data.data.email.code){
              console.log("email already exist")
            //put some code then return    
          }
    }

pb js sdk create operation would return:

    "code": 400,
        "message": "Failed to create record.",
        "data": {
            "email": {
                "code": "validation_invalid_email",
                "message": "The email is invalid or already in use."
            },
            "username": {
                "code": "validation_invalid_username",
                "message": "The username is invalid or already in use."
            }
        }
    }

An alternative approach could involve enabling users to enter a username on the HTML registration page, and then checking whether the username already exists before performing the create or save operation. To implement this, you could consider adding a custom route while using Pocketbase as a framework.

 	    e.Router.AddRoute(echo.Route{
			  Method: http.MethodPost,
			  Path:   "/api/checkuser",	
			  Handler: func(c echo.Context) error {
  
				  userx:=new(Userx)
				  if err := c.Bind(userx); err != nil {
					  return err
				  }
				  fmt.Println(userx)
								
				  expr := dbx.NewExp("username={:username}", dbx.Params{"username": userx.Username})
  				  record, err := app.Dao().FindRecordsByExpr("users", expr)
  
				  if err != nil {
					  return err
				  }
				  
				  if len(record) > 0 {
					  return c.String(http.StatusConflict, "User Exists")
				  } else {
					  return c.String(http.StatusAccepted, "OK")
  				  }
  			  },
			  Middlewares: []echo.MiddlewareFunc{
				  apis.ActivityLogger(app),
			  },
		  })

**Comment by TheKinng96** [comment]
Another workaround will need to fetch a list of available emails or username then check if exists
